subprojects/base-services/src/main/java/org/gradle/internal/reflect/DirectInstantiator.java

```diff
@@ -36,7 +39,14 @@ private DirectInstantiator() {
 
     public <T> T newInstance(Class<? extends T> type, Object... params) {
         try {
-            Constructor<?> match = doGetConstructor(type, constructorCache.get(type), params);
+            Class<?>[] argTypes = wrapArgs(params);
+            Constructor<?> match = null;
+            while (match == null) {
```

Comment:
It would perhaps be better to have the cache take care of this coordination. It could potentially do this by being provided a lookup function that, given a cache entry can locate the value.